File: public/dsa/linked_lists/middle_element_of_linked_list/slow_fast_pointer/middle_element_of_linked_list.cpp

#include <bits/stdc++.h>
using namespace std;

// defs-start
class ListNode {
   public:
    int data;
    ListNode* next;
    ListNode(int data, ListNode* next = nullptr) : data(data), next(next) {}
};

// defs-end

int getMiddleElementOfLinkedList(ListNode* head) {
    ListNode* slow = head;
    ListNode* fast = head;

    while (fast->next && fast->next->next) {
        slow = slow->next;
        fast = fast->next->next;
    }

    return slow->data;
}

// tests-start

int main() {
    ListNode* list1 = new ListNode(1, new ListNode(8, new ListNode(3)));
    ListNode* list2 = new ListNode(1, new ListNode(3, new ListNode(8, new ListNode(5))));
    cout << getMiddleElementOfLinkedList(list1) << endl;
    cout << getMiddleElementOfLinkedList(list2) << endl;
    return 0;
}

// tests-end

/*output
8
3
*/

File: public/dsa/linked_lists/middle_element_of_linked_list/slow_fast_pointer/middle_element_of_linked_list.py
# defs-start
class ListNode:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next
# defs-end


class Solution:
    def getMiddleElementOfLinkedList(self, head: ListNode) -> int:
        slow = head
        fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        return slow.data


# tests-start
list1 = ListNode(1, ListNode(8, ListNode(3)))
list2 = ListNode(1, ListNode(3, ListNode(8, ListNode(5))))
sol = Solution()
print(sol.getMiddleElementOfLinkedList(list1))
print(sol.getMiddleElementOfLinkedList(list2))
# tests-end

'''output
8
3
'''
